### PR TITLE
[opt](iceberg) optimize rewrite_data_files to avoid generating excessive small files

### PR DESCRIPTION
**Pull request by suxiaogang223**
## What problem does this PR solve?

During Iceberg rewrite_data_files operations, when BE count is large, an unexpected number of small files are generated.

Problem: total_files = task_count × active_BE_count × partition_count

## What is changed and how it works?

This change refines the parallelism strategy of Iceberg rewrite_data_files to maximize concurrency without producing
  extra output files.

  - If expectedFileCount <= alive BE count, use GATHER. Read parallelism is no longer forced to 1; it is capped at
    min(defaultParallelism, expectedFileCount) so output files never exceed the expected count.
  - If expectedFileCount > alive BE count, cap per‑BE parallelism to floor(expectedFileCount / aliveBeCount), and then take
    min with defaultParallelism, ensuring total writers do not exceed the expected file count.
  - Updated unit and regression tests to cover GATHER vs non‑GATHER paths and boundary cases.
## Benefits

- Small data: reduce from 100+ files to ~1 file (90%+ reduction)
- Adaptive strategy, no manual tuning needed

## Check List

- [x] Code changes
- [x] Test strategy described

### CODE REVIEW

**Comment by suxiaogang223** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE UT Coverage Report
Increment line coverage `  69.70% (23/33)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60063_46f10672d6ae692d2375641df8e9d82e21386224/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60063_46f10672d6ae692d2375641df8e9d82e21386224/fe_report/index.html)

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  15.15% (5/33)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60063_46f10672d6ae692d2375641df8e9d82e21386224_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60063_46f10672d6ae692d2375641df8e9d82e21386224_merge_fe/report/index.html)

**Comment by suxiaogang223** [comment]
run buildall

**Comment by hello-stephen** [comment]
# FE Regression Coverage Report
Increment line coverage `  2.86% (1/35)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/60063_59041882fdebea471e2be5fe60385b7ef3a60fbf_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/60063_59041882fdebea471e2be5fe60385b7ef3a60fbf_merge_fe/report/index.html)

File: fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteDataFileExecutor.java [hunk 1]
```diff
@@ -69,6 +69,10 @@ public RewriteResult executeGroupsConcurrently(List<RewriteDataGroup> groups, lo
         List<RewriteGroupTask> tasks = Lists.newArrayList();
         RewriteResultCollector resultCollector = new RewriteResultCollector(groups.size(), tasks);
 
+        // Get available BE count once before creating tasks
+        // This avoids calling getBackendsNumber() in each task during multi-threaded execution
+        int availableBeCount = Env.getCurrentSystemInfo().getBackendsNumber(true);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteDataFileExecutor.java]: We need to consider the compute group

File: fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java [hunk 2]
```diff
@@ -252,9 +260,78 @@ private ConnectContext buildConnectContext() {
...
+
+        // Threshold for using GATHER distribution
+        // When expected files <= this threshold, collect all data to single node
+        final int gatherThreshold = 1;
+
+        if (expectedFileCount <= gatherThreshold) {
+            // Small data volume: use GATHER to write to single node
+            useGather = true;
+            optimalParallelism = 1;
+            LOG.info("[Rewrite Task] taskId: {}, using GATHER distribution for small data. "
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java]: This log can be removed. Use one in line 313

File: fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java [hunk 3]
```diff
@@ -234,6 +237,11 @@ private ConnectContext buildConnectContext() {
         // Clone session variables from parent
         taskContext.setSessionVariable(VariableMgr.cloneSessionVariable(connectContext.getSessionVariable()));
 
+        // Calculate optimal parallelism and determine distribution strategy
+        RewriteStrategy strategy = calculateRewriteStrategy();
+        // Pipeline engine uses parallelPipelineTaskNum to control instance parallelism.
+        taskContext.getSessionVariable().parallelPipelineTaskNum = strategy.parallelism;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java:243]: Here we use a new value to overwrite the session variable?

Comment (suxiaogang223) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java:243]: Because the parameter of the setParallelPipelineTaskNum API is a string, it needs to be parsed first, so the value is assigned directly here.

**Comment by suxiaogang223** [comment]
run buildall

**Review by morningman — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java [hunk 4]
```diff
@@ -252,9 +260,86 @@ private ConnectContext buildConnectContext() {
...
+        LOG.info("[Rewrite Task] taskId: {}, totalSize: {} bytes, targetFileSize: {} bytes, "
+                        + "expectedFileCount: {}, availableBeCount: {}, defaultParallelism: {}, "
+                        + "optimalParallelism: {}, useGather: {}",
+                taskId, totalSize, targetFileSizeBytes, expectedFileCount,
+                availableBeCount, defaultParallelism, optimalParallelism, useGather);
+
+        return new RewriteStrategy(optimalParallelism, useGather);
+    }
+
+    private int getDefaultPipelineParallelism() {
```
Comment (zhangstar333) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/rewrite/RewriteGroupTask.java]: seems could call function directly SessionVariable#getParallelExecInstanceNum()

**Review by morningman — APPROVED** [review]

**Review by zhangstar333 — APPROVED** [review]
